### PR TITLE
feat: theorems for `ushiftRight`

### PR DESCRIPTION
**Pull request by mhk119**

### CODE REVIEW

**Comment by bollu** [comment]
@mhk119 I just pushed to the PR, and broke the proof strategy down into smaller chunks so it's easier to upstream. Let me know what you think, and if you're happy with the PR.

**Comment by mhk119** [comment]
> @mhk119 I just pushed to the PR, and broke the proof strategy down into smaller chunks so it's easier to upstream. Let me know what you think, and if you're happy with the PR.

This looks good !

File: src/Init/Data/Nat/Bitwise/Basic.lean [hunk 1]
```diff
@@ -71,6 +71,9 @@ theorem shiftRight_eq_div_pow (m : Nat) : ∀ n, m >>> n = m / 2 ^ n
     rw [shiftRight_add, shiftRight_eq_div_pow m k]
     simp [Nat.div_div_eq_div_mul, ← Nat.pow_succ, shiftRight_succ]
 
+theorem shiftRight_eq_zero (m n : Nat) (hn : m < 2 ^n): m >>> n = 0 := by
+  simp [Nat.shiftRight_eq_div_pow, Nat.div_eq_of_lt hn]
```
Comment (tobiasgrosser) [inline src/Init/Data/Nat/Bitwise/Basic.lean:75]:
```suggestion
theorem shiftRight_eq_zero (m n : Nat) (hn : m < 2 ^n): m >>> n = 0 := by
  simp [Nat.shiftRight_eq_div_pow, Nat.div_eq_of_lt hn]
```
```suggestion
theorem shiftRight_eq_zero (m n : Nat) (hn : m < 2 ^n) : m >>> n = 0 := by
  simp [Nat.shiftRight_eq_div_pow, Nat.div_eq_of_lt hn]
```

File: src/Init/Data/Int/Bitwise/Lemmas.lean [hunk 2]
```diff
@@ -34,4 +34,8 @@ theorem shiftRight_eq_div_pow (m : Int) (n : Nat) :
 theorem zero_shiftRight (n : Nat) : (0 : Int) >>> n = 0 := by
   simp [Int.shiftRight_eq_div_pow]
 
+@[simp]
+theorem shiftRight_zero (n : Int) : n >>> 0 = n := by
+  simp [Int.shiftRight_eq_div_pow]
```
Comment (tobiasgrosser) [inline src/Init/Data/Int/Bitwise/Lemmas.lean:39]: Should `n` be of type BitVec?

File: src/Init/Data/BitVec/Lemmas.lean [hunk 3]
```diff
@@ -1240,6 +1240,48 @@ theorem toNat_ushiftRight_lt (x : BitVec w) (n : Nat) (hn : n ≤ w) :
...
+  · have : x.toNat >>> n = 0 := by
+      apply Nat.shiftRight_eq_zero
+      have : 2^w ≤ 2^n := Nat.pow_le_pow_of_le (by decide) (by omega)
+      omega
+    simp [this] at h
+    omega
+
+@[simp]
+theorem toFin_uShiftRight (x : BitVec w) (n : Nat) :
+    (x >>> n).toFin = x.toFin / (Fin.ofNat' (2^w) (2^n)) := by
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Lemmas.lean:1277]:
```suggestion
theorem toFin_uShiftRight {x : BitVec w} {n : Nat} :
    (x >>> n).toFin = x.toFin / (Fin.ofNat' (2^w) (2^n)) := by
```

In the BitVec file, we use implicit arguments or equalities.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 4]
```diff
@@ -1240,6 +1240,48 @@ theorem toNat_ushiftRight_lt (x : BitVec w) (n : Nat) (hn : n ≤ w) :
...
+theorem ushiftRight_eq_zero (x : BitVec w) (n : Nat) (hn : w ≤ n) :
+    x >>> n = 0#w := by
+  simp [bv_toNat]
+  have : 2^w ≤ 2^n := Nat.pow_le_pow_of_le Nat.one_lt_two hn
+  rw [Nat.shiftRight_eq_div_pow, Nat.div_eq_of_lt (by omega)]
+
+
+/--
+Unsigned shift right by at least one bit makes the value of the bitvector less than or equal to `2^(w-1)`,
+makes the interpretation of the bitvector `Int` and `Nat` agree.
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Lemmas.lean:1254]: `makes the` two times in a row is grammatically suspicious.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 5]
```diff
@@ -1240,6 +1240,48 @@ theorem toNat_ushiftRight_lt (x : BitVec w) (n : Nat) (hn : n ≤ w) :
...
+-/
+theorem toInt_ushiftRight_pos_eq_toNat_shiftRight (x : BitVec w) (n : Nat) (hn : 0 < n) :
+    (x >>> n).toInt = x.toNat >>> n := by
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Lemmas.lean:1257]:
```suggestion
theorem toInt_ushiftRight_of_lt {x : BitVec w} {n : Nat} (hn : 0 < n) :
    (x >>> n).toInt = x.toNat >>> n := by
```

File: src/Init/Data/BitVec/Lemmas.lean [hunk 6]
```diff
@@ -1240,6 +1240,48 @@ theorem toNat_ushiftRight_lt (x : BitVec w) (n : Nat) (hn : n ≤ w) :
     · apply hn
   · apply Nat.pow_pos (by decide)
 
+
+/-- Shifting right by `n`, which is larger than the bitwidth `w` produces `0. -/
+theorem ushiftRight_eq_zero (x : BitVec w) (n : Nat) (hn : w ≤ n) :
+    x >>> n = 0#w := by
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Lemmas.lean:1246]:
```suggestion
theorem ushiftRight_eq_zero {x : BitVec w} {n : Nat} (hn : w ≤ n) :
    x >>> n = 0#w := by
```

File: src/Init/Data/BitVec/Lemmas.lean [hunk 7]
```diff
@@ -1240,6 +1240,48 @@ theorem toNat_ushiftRight_lt (x : BitVec w) (n : Nat) (hn : n ≤ w) :
...
+  simp only [toNat_ushiftRight, ite_eq_left_iff, Nat.not_lt]
+  intros h
+  by_cases hn: n ≤ w
+  · have h1 := Nat.mul_lt_mul_of_pos_left (toNat_ushiftRight_lt x n hn) Nat.two_pos
+    simp only [toNat_ushiftRight, Nat.zero_lt_succ, Nat.mul_lt_mul_left] at h1
+    have : 2 ^ (w - n).succ ≤ 2^ w := Nat.pow_le_pow_of_le (by decide) (by omega)
+    have := show 2 * x.toNat >>> n < 2 ^ w by
+      omega
+    omega
+  · have : x.toNat >>> n = 0 := by
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Lemmas.lean:1268]: Can this be simplified with `ushiftRight_eq_zero`?

**Comment by tobiasgrosser** [comment]
@bollu, can this be closed?

**Pull request closed without merging**